495: Default string values with dashes are mistakenly treated as dynamic defaults

lindsay-stevens opened the pull request:
Closes #495

#### Why is this the best possible solution? Were any other approaches considered?

At this stage just a draft PR with tests to reproduce the issues described in 495 and 533.

#### What are the regression risks?

TBC

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.

TBC

#### Before submitting this PR, please make sure you have:
- [ ] included test cases for core behavior and edge cases in `tests`
- [ ] run `nosetests` and verified all tests pass
- [ ] run `black pyxform tests` to format code
- [ ] verified that any code or assets from external sources are properly credited in comments

lindsay-stevens commented:
Hi @lognaturel there may be some more to do on this wrt. tests but I think it is ready for a look. Thanks!

lognaturel commented on tests/test_dynamic_default.py:
```diff
@@ -337,3 +339,203 @@ def test_dynamic_default_on_calculate(self):
+            # Literal text containing URI.
+            self.Case(False, "text", """go to https://mysite.com""", ""),
+            # Literal text containing various non-operator symbols.
+            self.Case(False, "text", """Repeat after me: ~!@#$%^&()_""", ""),
```
lol

lognaturel commented on tests/test_dynamic_default.py:
```diff
@@ -337,3 +339,203 @@ def test_dynamic_default_on_calculate(self):
+        self.case_data_extras = (
+            # TODO: pyxform references appear in result as resolved XPaths
+            # Pyxform reference.
+            self.Case(True, "text", """${q0}""", ""),
```
I'm not clear on what this todo is about? Is the issue that the test XML output will have `/data/q0` in it? Actually I can't figure out how it's passing?

lindsay-stevens replied:
Yep the issue is that the output has the resolved XPath e.g. `/data/q0` form rather than `${q0}`. Possibly a matter of reworking the tests to allow for this case specifying an alternative expected value. The test passes with this Case because the `self.case_data_extras` items are only passed through `utils.default_is_dynamic` in the shorter test, not `assertPyxformXform` in the longer test.

lindsay-stevens replied:
Resolved this by updating the test strategy for cases that initially or after rendering contain single quotes, namely to use xpath_exact to compare outside of xpath, instead of xpath_match. There is one remaining "extra" case for the union operator, but the fate of that case depends on https://github.com/getodk/xforms-spec/issues/293

lognaturel replied:
Ah yes, intriguing.

lognaturel reviewed (commented):
Definitely more involved than I would have gone for but more correct, too!

Here are some cases I played around with:
```
            self.Case(True, "date", """concat('2022-03', '-14')""", ""),

            self.Case(False, "text", """f-4""",""),
            self.Case(False, "text", """./f-4""",""),
            self.Case(True, "text", """./f - 4""",""),
            self.Case(False, "integer", """7-4""",""),
            self.Case(True, "integer", """7 - 4""",""),
```

The behavior with `-` is a change so that could possibly break some form updates. I can't imagine these types of cases are common but it might be good to have the tests as explicit documentation. And as you said, also add something to user-facing docs.

lindsay-stevens commented:
Thanks for the review! I will update the tests as mentioned above to complete this PR.

lindsay-stevens commented on tests/test_dynamic_default.py:
```diff
@@ -337,3 +339,203 @@ def test_dynamic_default_on_calculate(self):
+        self.assertPyxformXform(
+            name="test",
+            md=md,
+            run_odk_validate=True,
```
Added tests for this

lindsay-stevens commented:
Should be good to go now. New commit is all about tests, per commit message for [ba7b196](https://github.com/XLSForm/pyxform/pull/578/commits/ba7b1967973da39002a89287bf27dde8cb6e7f15):

- deleted existing tests where covered by case in TestDynamicDefaultSimpleInput
- updated tests using string matchers to use xpath instead
- added pyxform_test_case ability to escape literal pipe in values
- clarified existing translation performance test description
- added util function for coalesce
- added tests suggested by @lognaturel and performance tests
- updated test strategy for cases that initially or after rendering
  contain single quotes, namely to use xpath_exact to compare outside
  of xpath, instead of xpath_match.

The performance tests are:
- `test_dynamic_default_performance__time`: processing time for 500, 1000, 2000, 5000 dynamic default questions shouldn't take much longer than no-op dynamic defaults check (results were approx 0 to 6%).
- `test_dynamic_default_performance__memory`: question with 2000 dynamic default questions shouldn't increase memory usage more than x2.

lognaturel commented:
🥳 🥳 🥳
